core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -520,7 +521,13 @@ sealed abstract class IO[+E, +A] { self =>
   /**
    * Runs this action in a new fiber, resuming when the fiber terminates.
    */
-  final def run[E1 >: E, A1 >: A]: IO[Nothing, ExitResult[E1, A1]] = new IO.Run(self)
+  final def run: IO[Nothing, ExitResult[E, A]] =
+    for {
+      p <- Promise.make[Nothing, ExitResult[E, A]]
+      f <- self.fork
+      _ <- f.onComplete(r => p.done(ExitResult.Completed(r)).void)
+      r <- p.get
+    } yield r
```

Comment:
Add `.supervised` to this whole block to kill `self.fork` when the parent exits.